ddtrace/contrib/internal/langchain/patch.py

```diff
@@ -974,6 +975,88 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
     return documents
 
 
+@with_traced_module
+def traced_base_tool_invoke(langchain, pin, func, instance, args, kwargs):
+    integration = langchain._datadog_integration
+    tool_input = get_argument_value(args, kwargs, 0, "input")
+    config = get_argument_value(args, kwargs, 1, "config", optional=True)
+
+    span = integration.trace(
+        pin,
+        "%s" % func.__self__.name,
+        interface_type="tool",
+    )
+
+    tool_output = None
+    try:
+        for attribute in ("name", "description", "metadata", "tags"):
+            value = getattr(instance, attribute, None)
+            if isinstance(value, dict):
+                for key, meta_value in value.items():
+                    span.set_tag_str("langchain.request.tool.%s.%s" % (attribute, key), str(value))
+            elif isinstance(value, list):
+                for idx, tag in enumerate(value):
+                    span.set_tag_str("langchain.request.tool.%s.%d" % (attribute, idx), str(value))
+            elif value is not None:
+                span.set_tag_str("langchain.request.tool.%s" % attribute, str(value))
```

Comment:
I think this might be too much nesting. I'm find with handling `name` and `description` in the same for loop since they are simple strings, but can we handle `metadata` (dict) and `tags` (list of strings) separately?

Reply:
Yes i can do that !
for future references why is nesting capped ? Isn't complexity exactly the same since we iterate on a fixed length list ?

Reply:
Yeah complexity/performance is not affected, it's just Python styling best practices to avoid excessive nesting for readability